gcc/testsuite/rust/execute/torture/slice-magic.rs

```diff
@@ -0,0 +1,106 @@
+// { dg-additional-options "-w" }
+extern "rust-intrinsic" {
+    pub fn offset<T>(dst: *const T, offset: isize) -> *const T;
+}
+
+struct FatPtr<T> {
+    data: *const T,
+    len: usize,
+}
+
+union Repr<T> {
+    rust: *const [T],
+    rust_mut: *mut [T],
+    raw: FatPtr<T>,
+}
+
+#[lang = "Range"]
+pub struct Range<Idx> {
+    pub start: Idx,
+    pub end: Idx,
+}
+
+#[lang = "const_slice_ptr"]
+impl<A> *const [A] {
```

Comment:
Just as a heads up https://github.com/rust-lang/rust/pull/94963 changed the way this kind of impls on builtin and foreign types are handled. Libcore now uses `#![rustc_coherence_is_core]` to allow impls on any builtin type without `#[lang = "..."]` and other sysroot crates use `#[rustc_allow_incoherent_impl]` on all impl items for non-local impls.